File: app/src/main/java/com/jeppsson/appexplore/AppComponentUtils.java
package com.jeppsson.appexplore;

import android.content.pm.ActivityInfo;
import android.content.pm.PackageInfo;
import android.content.pm.ProviderInfo;
import android.content.pm.ServiceInfo;

final class AppComponentUtils {

    private static final int LAUNCH_MULTIPLE = 0;
    private static final int LAUNCH_SINGLE_TOP = 1;
    private static final int LAUNCH_SINGLE_TASK = 2;
    private static final int LAUNCH_SINGLE_INSTANCE = 3;

    public static String getActivities(PackageInfo packageInfo) {
        StringBuilder sb = new StringBuilder();

        if (packageInfo.activities != null) {
            for (ActivityInfo activityInfo : packageInfo.activities) {
                sb.append(activityInfo.name).append('\n');
                if (activityInfo.exported) {
                    sb.append("Exported: true").append('\n');
                }
                if (activityInfo.launchMode != 0) {
                    sb.append("LaunchMode: ")
                            .append(launchMode(activityInfo.launchMode))
                            .append('\n');
                }
            }
        } else {
            sb.append("No activities");
        }

        return sb.toString().trim();
    }

    public static String getServices(PackageInfo packageInfo) {
        StringBuilder sb = new StringBuilder();

        if (packageInfo.services != null) {
            for (ServiceInfo serviceInfo : packageInfo.services) {
                sb.append(serviceInfo.name).append('\n')
                        .append("Exported: ").append(serviceInfo.exported).append('\n');
            }
        } else {
            sb.append("No services");
        }

        return sb.toString().trim();
    }

    public static String getContentProviders(PackageInfo packageInfo) {
        StringBuilder sb = new StringBuilder();

        if (packageInfo.providers != null) {
            for (ProviderInfo providerInfo : packageInfo.providers) {
                sb.append(providerInfo.name).append('\n')
                        .append("Exported: ").append(providerInfo.exported).append('\n');
            }
        } else {
            sb.append("No content providers");
        }

        return sb.toString().trim();
    }

    public static String getReceivers(PackageInfo packageInfo) {
        StringBuilder sb = new StringBuilder();

        if (packageInfo.receivers != null) {
            for (ActivityInfo activityInfo : packageInfo.receivers) {
                sb.append(activityInfo.name).append('\n')
                        .append("Exported: ").append(activityInfo.exported).append('\n');
            }
        } else {
            sb.append("No receivers");
        }

        return sb.toString().trim();
    }


    private static String launchMode(int launchMode) {
        switch (launchMode) {
            case LAUNCH_MULTIPLE:
                return "standard";

            case LAUNCH_SINGLE_TOP:
                return "singleTop";

            case LAUNCH_SINGLE_TASK:
                return "singleTask";

            case LAUNCH_SINGLE_INSTANCE:
                return "singleInstance";

            default:
                return "unknown";
        }
    }
}

File: app/src/main/java/com/jeppsson/appexplore/utils/AppComponentUtils.kt
package com.jeppsson.appexplore.utils

import android.content.pm.PackageInfo
import android.graphics.Color
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import java.util.regex.Pattern

internal object AppComponentUtils {

    private const val LAUNCH_MULTIPLE = 0
    private const val LAUNCH_SINGLE_TOP = 1
    private const val LAUNCH_SINGLE_TASK = 2
    private const val LAUNCH_SINGLE_INSTANCE = 3

    fun getActivities(packageInfo: PackageInfo): Spannable {
        val sb = StringBuilder()

        if (packageInfo.activities != null) {
            for (activityInfo in packageInfo.activities) {
                sb.append(activityInfo.name).append('\n')
                val info = StringBuilder()
                if (activityInfo.exported) {
                    info.append("(exported)").append(' ')
                }
                if (activityInfo.launchMode != LAUNCH_MULTIPLE) {
                    info.append('(')
                            .append(launchMode(activityInfo.launchMode))
                            .append(')')
                            .append(' ')
                }
                if (info.isNotEmpty()) {
                    info.append('\n')
                }
                sb.append(info)
            }
        } else {
            sb.append("No activities")
        }

        return colorize(sb.toString().trim())
    }

    fun getServices(packageInfo: PackageInfo): Spannable {
        val sb = StringBuilder()

        if (packageInfo.services != null) {
            for (serviceInfo in packageInfo.services) {
                sb.append(serviceInfo.name).append('\n')
                if (serviceInfo.exported) {
                    sb.append("(exported)").append('\n')
                }
            }
        } else {
            sb.append("No services")
        }

        return colorize(sb.toString().trim())
    }

    fun getContentProviders(packageInfo: PackageInfo): Spannable {
        val sb = StringBuilder()

        if (packageInfo.providers != null) {
            for (providerInfo in packageInfo.providers) {
                sb.append(providerInfo.name).append('\n')
                if (providerInfo.exported) {
                    sb.append("(exported)").append('\n')
                }
            }
        } else {
            sb.append("No content providers")
        }

        return colorize(sb.toString().trim())
    }

    fun getReceivers(packageInfo: PackageInfo): Spannable {
        val sb = StringBuilder()

        if (packageInfo.receivers != null) {
            for (activityInfo in packageInfo.receivers) {
                sb.append(activityInfo.name).append('\n')
                if (activityInfo.exported) {
                    sb.append("(exported)").append('\n')
                }
            }
        } else {
            sb.append("No receivers")
        }

        return colorize(sb.toString().trim())
    }


    private fun launchMode(launchMode: Int): String {
        return when (launchMode) {
            LAUNCH_MULTIPLE -> "standard"
            LAUNCH_SINGLE_TOP -> "singleTop"
            LAUNCH_SINGLE_TASK -> "singleTask"
            LAUNCH_SINGLE_INSTANCE -> "singleInstance"
            else -> "unknown"
        }
    }

    private fun colorize(text: String): SpannableString {
        val content = SpannableString(text)

        val m = Pattern.compile("\\b(?:exported)\\b").matcher(text)
        while (m.find()) {
            content.setSpan(ForegroundColorSpan(Color.YELLOW),
                    m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        }

        return content
    }
}
